src/datasets/data_files.py

```diff
@@ -89,16 +89,35 @@ def sanitize_patterns(patterns: Union[Dict, List, str]) -> Dict[str, Union[List[
     The default split is "train".
 
     Returns:
-        patterns: dictionary of split_name -> list_of _atterns
+        patterns: dictionary of split_name -> list_of patterns
     """
     if isinstance(patterns, dict):
         return {str(key): value if isinstance(value, list) else [value] for key, value in patterns.items()}
     elif isinstance(patterns, str):
         return {SANITIZED_DEFAULT_SPLIT: [patterns]}
     elif isinstance(patterns, list):
-        return {SANITIZED_DEFAULT_SPLIT: patterns}
+        if any(isinstance(pattern, dict) for pattern in patterns):
+            for pattern in patterns:
+                if not isinstance(pattern, dict) or sorted(pattern) != ["pattern", "split"]:
+                    raise ValueError(
+                        f"Expected data_files in YAML to be a string or a list, but got {pattern}\nExamples:\n"
+                        "    data_files: data.csv\n    data_files: data/*.png\n"
+                        "    data_files:\n    - part0/*\n    - part1/*\n"
+                        "    data_files:\n    - split: train\n      pattern: train/*\n    - split: test\n      pattern: test/*"
+                    )
+            splits = [pattern["split"] for pattern in patterns]
+            if len(set(splits)) != len(splits):
+                raise ValueError(f"Some splits are duplicated in data_files: {splits}")
+            return {
+                str(pattern["split"]): pattern["pattern"]
+                if isinstance(pattern["pattern"], list)
+                else [pattern["pattern"]]
+                for pattern in patterns
+            }
+        else:
+            return {SANITIZED_DEFAULT_SPLIT: patterns}
```

Comment:
This is to allow reading the data_files formatted in YAML:
```yaml
data_files:
  - split: train
    pattern: train_data/*
```

and we don't document that passing the same structure in python also works

Reply:
Yes, I know, but the YAML error here is "unexpected" from the design standpoint. I think it would be cleaner if `MetadataConfigs` would convert the `data_files` to the format that `sanitaze_patterns` can understand (`{"train": "train_data/*"}` for the above example or raise an error if the YAML is not valid).

PS: If you want to make it easy to convert the YAML config to Python kwargs that can then be passed to `load_dataset` (and the other way around), I guess we could allow this format in `sanitaze_patterns`, but without mentioning YAML.

Reply:
I think it's simpler to allow this structure in sanitize_patterns indeed. Maybe we can make the error message YAML-agnostic and move this YAML error message to the place where MetadataConfigs is converted to BuilderConfigs and resolving the data files.

Reply:
[8f6dc46](https://github.com/huggingface/datasets/pull/5331/commits/8f6dc4637e90a3cf0a8394804bd56c808b85dbf9)